src/json-api-adapter.js

```diff
@@ -10,44 +10,79 @@ DS._routes = Ember.create(null);
 
 DS.JsonApiAdapter = DS.RESTAdapter.extend({
   defaultSerializer: 'DS/jsonApi',
+
+  getSelfRoute: function(typeName, id, record) {
+    if(!this.serializer || !this.serializer.buildSelfKey) { return null; }
+    var route;
+    if(record) {
+      route = DS._routes[this.serializer.buildSelfKey(record.type, record.id, typeName, id)]
+        || DS._routes[this.serializer.buildSelfKey(record.type, null, typeName, null)];
+      if(route) { return route; }
+    }
+    return DS._routes[this.serializer.buildSelfKey(null, null, typeName, id)]
+      || DS._routes[this.serializer.buildSelfKey(null, null, typeName, null)];
+  },
+
+  getResourceRoute: function(typeName, id) {
+    var routeName = (id) ? typeName + '.' + id : typeName;
+    return DS._routes[routeName] || DS._routes[typeName];
+  },
+
+  getRoute: function(typeName, id, record) {
+    var route;
+    if(record) {
+      route = this.getSelfRoute(typeName, id, record);
+      if(route) { return route; }
+    }
+    return this.getResourceRoute(typeName, id);
```

Comment:
This is just precursor work to support both the `resource` link and the `self` link. The problem is the way `DS.Store` is interacting with the adapter is too limiting. From the JSON API perspective, you should call `resource` to retrieve the details of that resource and you should operate on `self` if you needed to update the association between the two resources. JSON API is too vague on this concept and sometimes `self` should be used to retrieve the deals of the association.

In looking at JSON API documentation right now, it appears they have changed `resource` to `related`, so I will make that update, but the concept remains the same. Take a look at [Resource Relationships](http://jsonapi.org/format/#document-structure-resource-relationships) for more details.